core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
I would argue that the return value of that method should be effect-free (in the context of the model - obviously there are effects we cannot model.) For example considering `Eval` - which I consider the motivating example, let me know if there are others you think are as relevant - you are never going to be given an unevaluated metadata map. If there is unevaluated metadata, that would be modeled in the map, e.g. with a key which produced `Eval[Data]` instead of `Data`.

The reasoning is that this method has to return instantaneously or nearly so in order to have a usable filesystem. So I think it's counterproductive to provide a layer of indirection between the input and the output.

It's like the difference between a `Lazy[List[A]]` and a `List[Lazy[A]]`. We need the second, not the first.